crowbar_framework/lib/crowbar/upgrade_status.rb

```diff
@@ -92,17 +92,22 @@ def start_step(step_name, out_of_order = false)
       end
     end
 
-    def end_step(success = true, errors = {})
+    def end_step(step_name = nil, out_of_order = false, success = true, errors = {})
       ::Crowbar::Lock::LocalBlocking.with_lock(shared: false, logger: @logger, path: lock_path) do
-        unless running?
+        step_name ||= current_step
+        unless running? step_name
           @logger.warn("The step is not running, could not be finished")
           return false
         end
-        progress[:steps][current_step] = {
+        if step_name != current_step && !out_of_order
+          @logger.error("The end of step #{step_name} is requested in the wrong order")
+          return false
+        end
+        progress[:steps][step_name] = {
           status: success ? :passed : :failed,
           errors: errors
         }
-        next_step
+        next_step unless out_of_order
```

Comment:
The point is, we do not want to move `current_step` to next one if we were executing out of order. And the called has the knowledge about out-of-order status, so I think it's better to provide this info both when starting and ending the step